pkg/kubelet/container/sync_result.go

```diff
@@ -19,14 +19,59 @@ package container
 import (
 	"errors"
 	"fmt"
+	"time"
 
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 )
 
 // TODO(random-liu): We need to better organize runtime errors for introspection.
 
-// ErrCrashLoopBackOff returned when a container Terminated and Kubelet is backing off the restart.
-var ErrCrashLoopBackOff = errors.New("CrashLoopBackOff")
+// BackoffError should be used whenever an error needs to specify a particular backoff duration
+// to the Kubelet.
+type BackoffError struct {
+	error
+	backoffTime time.Time
+}
+
+func NewBackoffError(err error, backoffTime time.Time) *BackoffError {
+	return &BackoffError{
+		error:       err,
+		backoffTime: backoffTime,
+	}
+}
+
+// BackoffTime returns the expected expiration time of the backoff.
+func (e *BackoffError) BackoffTime() time.Time {
+	return e.backoffTime
+}
+
+// MinBackoffExpiration recursively searches through err for BackoffErrors and returns
+// the minimum of all found backoff times.
+func MinBackoffExpiration(err error) (time.Time, bool) {
+	var ae utilerrors.Aggregate
+	var be *BackoffError
+	switch {
+	case errors.As(err, &be):
+		return be.BackoffTime(), true
+	case errors.As(err, &ae):
+		var min time.Time
+		found := false
+		for _, e := range ae.Errors() {
+			if backoff, ok := MinBackoffExpiration(e); ok {
+				if !found || backoff.Before(min) {
+					min = backoff
+					found = true
+				}
+			}
+		}
+		return min, found
+	default:
+		if e := errors.Unwrap(err); e != nil {
+			return MinBackoffExpiration(e)
+		}
+		return time.Time{}, false
```

Comment:
it this is returned, than it would be a large negative number when substituted from now? Can we ever get to this point? If so - would time.Now be a better thing to return from here?

Reply:
If we reach here, we also return `false`, so the caller shouldn't use the time value at all.